CPS-????/README.md

```diff
@@ -0,0 +1,173 @@
+This is an inevitable consequence of non-determinism: since the user has not specified exactly how things happen, the system or the actors within it are free to arrange the eventual fulfilment in a way that is maximally profitable for them. 
+This is a major topic, but one that needs to be addressed (or explicitly allowed) by an intent-processing system.
+
+## Use cases
+
+### Open (atomic) swaps
+
+An open swap is a request to make a specific exchange, but with an unknown counterparty. 
+Open swaps are very common, since users generally do not care who exactly they are trading with as long as they get the price that they want.
+
+Open swaps are the primary kind of intent handled by DEXs.
+
+Since swaps typically require a counterparty to be found, they are very reliant on intent distribution in order to ensure that user intents reach appropriate counterparties for fulfilment.
+
+#### Market orders
+
+A market order is a kind of swap where the asset price is the current “market price” (e.g. given by an on-chain oracle). 
+A market order therefore does not specify the actual price that the user is going to pay, although it will specify where the money comes from and how the price should be determined (e.g. via a particular oracle).
+
+#### Limit orders
+
+A limit order is a kind of swap where the asset price is defined as a constraint, for example “it must be above or below a given threshold”. 
+But the price may go past the threshold (“sell for at least 5”, but the price is 10), and then the order can be satisfied at that price. 
+Similar to a market order, a limit order therefore does not specify the actual price.
+
+#### Babel fees
+
+A Babel fee transaction is one where the user pays for the normal transaction fees using a token other than the base currency of the chain. 
+We can see such transactions as a combination of a normal transaction and an intent to swap their tokens for a sufficient quantity of the base currency.
+
+### DApp fee sponsorship
+
+An application operator may wish to cover the cost of fees for running their application, in order to make it more attractive to users. 
+In the Ethereum world, this phenomenon led to the well-known Gas Station Network.
+
+A transaction requiring fee sponsorship does not specify where the money for the fees is coming from. 
+The sponsor can fulfil it by picking it up and providing the money.
+
+In principle intents that need sponsorship can be sent directly to the application operator, which means these kinds of intents may need less sophisticated networking, at the cost of centralisation.
+
+### Non-deterministic access to state
+
+A user who interacts with an application’s state stored in the datum of a UTXO may well be happy to interact with the “current state” whatever that is. 
+An intent of this kind must specify the address of the UTXO it is spending, but not its content, and similarly not specify the state of the output it creates.
```

Comment:
Maybe better to describe this not as "the address of the UTxO" but more like, transactions with outputs that have a fill-in-able datum

Reply:
Sometimes the spender provides the datum for the output they are spending, if the output is a hash. So if you want to be insensitive to that, then someone else needs to provide that. So it can apply on both end.s

Reply:
`(a, v, ???) \in outputs tx1` has index 1, lets say
And inputs now can be "resolved" inside the transaction by specifying only the missing data
`(txId tx1, 1) |-> (_, _, myDatum) \in inputs tx2`

Do you mean something like this?

Reply:
No, the other way around. Normally when you spend an output, you have to either:
1. If the output has a datum hash, specify the pre-image
2. Otherwise do nothing

You might also have to specify a datum or a datum hash for the output that you are creating.

To do "nondeterministic access to state", you need to specify _neither_ datum. So that means either being in case 2, or in case 1 and not providing the pre-image; and also creating an output which doesn't specify the datum.

To make this complete, someone would need to fill in those details. I don't really know how.